semantic_cache/semantic_caching.py

```diff
@@ -0,0 +1,188 @@
+    """
+
+    def __init__(self, url: str, token: str, min_proximity: float = 0.9) -> None:
+        """
+        Initializes the SemanticCache with the given URL and token.
+
+        Args:
+            url (str): The URL of the Upstash Vector database.
+            token (str): The token for accessing the Upstash Vector database.
+            min_proximity (float): The minimum proximity score to consider a cache hit.
+        """
+        self.min_proximity = min_proximity
+        self.index = Index(url=url, token=token)
+
+    def get(self, key: str) -> Optional[str]:
+        """
+        Searches the cache for the key and returns the value if it exists.
+
+        Args:
+            key (str): The key to search in the cache.
+
+        Returns:
+            Optional[str]: The value associated with the key
+            if it exists and meets the proximity score; otherwise, None.
+        """
+        response = self._query_key(key)
+        if response is None or response.score <= self.min_proximity:
+            return None
+        return response.metadata["data"]
+
+    def lookup(
+        self, prompt: str, llm_string: Optional[str] = None
+    ) -> Optional[List[Generation]]:
+        """
+        Converts the JSON string to generations and returns them.
+        It is called by Langchain to check if the generations are in the cache.
+
+        Args:
+            prompt (str): The prompt to lookup in the cache.
+            llm_string (Optional[str]): Optional string for LLM context.
+
+        Returns:
+            Optional[List[Generation]]:
+            If response is found in the cache; otherwise, None.
+        """
+        result = self.get(prompt)
+        return self._loads_generations(result) if result else None
+
+    def update(
+        self,
+        prompt: str,
+        llm_string: Optional[str] = None,
+        result: Optional[str] = None,
+    ) -> None:
+        """
+        Converts the generations to a JSON string and stores it in the cache.
+        It is called by Langchain to upload the generations to the cache.
+
+        Args:
+            prompt (str): The prompt to update in the cache.
+            llm_string (Optional[str]): Optional string for LLM context.
+            result (Optional[str]): The result to store in the cache.
+        """
+        self.set(prompt, self._dumps_generations(result))
+
+    def set(self, key: Union[str, List[str]], data: Union[str, List[str]]) -> None:
+        """
+        Sets the key and value in the cache.
+
+        Args:
+            key (Union[str, List[str]]): The key(s) to set in the cache.
+            data (Union[str, List[str]]): The value(s) to associate with the key(s).
+        """
+        if isinstance(key, list) and isinstance(data, list):
+            for i in range(len(key)):
+                currrent_key = key[i]
+                self.index.upsert(
+                    [(self._hash_key(currrent_key), currrent_key, {"data": data[i]})]
+                )
```

Comment:
This is inefficient. Upsert can be used to upsert many items in a batch. So, you should iterate over the key list, and create a batch, and then call index.upsert with that batch, instead of upserting one by one here

Reply:
I updated it, didn't know that upsert and delete can be used like this